### ISSUE TITLE
[Bug]: torch.nn.functional.softmax converted to openvino gives different results in every run on intel gpu

### ISSUE THREAD

**Issue by DavitGrigoryan132** [openvinotoolkit/openvino#22851]
### OpenVINO Version

2024.0.0-14473-3238290df5e

### Operating System

Other (Please specify in description)

### Device used for inference

GPU

### Framework

None

### Issue description

I wrote simple model to reproduce bug which uses only torch.nn.functional.softmax and the converted and compiled openvino model gives different results in every run while running on intel gpu but gives correct results on cpu and nvidia gpu

    class CustomModel(torch.nn.Module):
        def __init__(self):
            super().__init__()
    
        # x has shape (1, 4096, 4096)
        def forward(self, x):
            x = torch.nn.functional.softmax(x, 1)
    
        return x

This is the model for bug reproduce.

### Step-by-step reproduction

Here is the script to reproduce bug

    import torch
    import onnx
    import openvino as ov
    import numpy as np
    
    
    class CustomModel(torch.nn.Module):
        def __init__(self):
            super().__init__()
    
        # x has shape (1, 4096, 4096)
        def forward(self, x):
            x = torch.nn.functional.softmax(x, 1)
    
            return x
    
    if __name__ == "__main__":
        model = CustomModel()
        model = model.to("cpu")
        model = model.eval()
    
        dummy_input = torch.rand((1, 4096, 4096), requires_grad=False)
    
        with torch.no_grad():
            _ = model(dummy_input)
    
            torch.onnx.export(model,
                              dummy_input,
                              "model.onnx",
                              export_params=True,
                              verbose=True,
                              do_constant_folding=True,
                              input_names=['input'],
                              output_names=['output'])
    
    
        core = ov.Core()
        ov_model = ov.convert_model("model.onnx")
        compiled_model = core.compile_model(model=ov_model, device_name="GPU.0")
    
        np.random.seed(1111)
        openvino_dummy_input = np.random.rand(1, 4096, 4096)
    
        result1 = compiled_model(openvino_dummy_input)[0]
        result2 = compiled_model(openvino_dummy_input)[0]
    
        print(result1)
        print(result2)
        print(np.all(result1 == result2))

this is the output for first run

> [[[0.00019956 0.00019324 0.00014186 ... 0.00036478 0.00021946 0.00016987]
>   [0.00018871 0.00037932 0.00032663 ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   ...
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]]]
> [[[0.00015569 0.00035691 0.00019944 ... 0.00037932 0.00032663 0.00033379]
>   [0.00014687 0.00016713 0.00026989 ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   ...
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]]]
> False

This is for second run

> [[[0.00014198 0.00017929 0.00017977 ... 0.00037932 0.00032663 0.00033379]
>   [0.00014687 0.00016713 0.00026989 ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   ...
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]]]
> [[[0.00014198 0.00017929 0.00017977 ... 0.00037885 0.0003264  0.00033355]
>   [0.00014675 0.00016701 0.00026965 ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   ...
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]]]
> False

And this is the output on CPU

> [[[0.00015661 0.00035816 0.00019998 ... 0.00026445 0.00030181 0.00036524]
>   [0.00022008 0.00016991 0.00018854 ... 0.00026914 0.00015746 0.00024553]
>   [0.00028234 0.00036734 0.00015755 ... 0.00020903 0.00031585 0.00014703]
>   ...
>   [0.00014377 0.0003858  0.00036929 ... 0.00017415 0.00025984 0.00027032]
>   [0.00036044 0.00024769 0.00019247 ... 0.00022465 0.00028568 0.00030115]
>   [0.00020977 0.00031799 0.00031262 ... 0.00028168 0.00024877 0.00015698]]]
> [[[0.00015661 0.00035816 0.00019998 ... 0.00026445 0.00030181 0.00036524]
>   [0.00022008 0.00016991 0.00018854 ... 0.00026914 0.00015746 0.00024553]
>   [0.00028234 0.00036734 0.00015755 ... 0.00020903 0.00031585 0.00014703]
>   ...
>   [0.00014377 0.0003858  0.00036929 ... 0.00017415 0.00025984 0.00027032]
>   [0.00036044 0.00024769 0.00019247 ... 0.00022465 0.00028568 0.00030115]
>   [0.00020977 0.00031799 0.00031262 ... 0.00028168 0.00024877 0.00015698]]]
> True

### Relevant log output

```shell
[[[0.00019956 0.00019324 0.00014186 ... 0.00036478 0.00021946 0.00016987]
  [0.00018871 0.00037932 0.00032663 ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  ...
  [0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]]]
[[[0.00015569 0.00035691 0.00019944 ... 0.00037932 0.00032663 0.00033379]
  [0.00014687 0.00016713 0.00026989 ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  ...
  [0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]]]
False
```

### Issue submission checklist

- [X] I'm reporting an issue. It's not a question.
- [X] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [X] There is reproducer code and related data files such as images, videos, models, etc.

**Comment by DavitGrigoryan132** [comment]
I simplified my custom class a bit to reproduce the bug and tried to write a softmax function myself. And this function gave the same problem. I tried to remove some layers from the model and find the minimal model where this bug occurs, and I got this model

    class CustomModel(torch.nn.Module):
        def __init__(self):
            super().__init__()
    
        def softmax(self, x, dim=0):
            # x_max = torch.max(x, dim=dim, keepdim=True)[0]
            # x_shifted = x - x_max
    
            x_shifted = x
            exp_x = torch.exp(x_shifted)
    
            exp_x_sum = torch.sum(exp_x, dim=dim, keepdim=True)
    
            softmax_output = exp_x / exp_x_sum
    
            return exp_x, exp_x_sum, softmax_output
    
        # x has shape (1, 4096, 4096)
        def forward(self, x):
            exp_x, exp_x_sum, softmax_output = self.softmax(x, 1)
    
            return exp_x, exp_x_sum, softmax_output

This is my whole script for reproducing bug

    import torch
    import onnx
    import openvino as ov
    import numpy as np
    
    
    class CustomModel(torch.nn.Module):
        def __init__(self):
            super().__init__()
    
        def softmax(self, x, dim=0):
            # x_max = torch.max(x, dim=dim, keepdim=True)[0]
            # x_shifted = x - x_max
    
            x_shifted = x
            exp_x = torch.exp(x_shifted)
    
            exp_x_sum = torch.sum(exp_x, dim=dim, keepdim=True)
    
            softmax_output = exp_x / exp_x_sum
    
            return exp_x, exp_x_sum, softmax_output
    
        # x has shape (1, 4096, 4096)
        def forward(self, x):
            exp_x, exp_x_sum, softmax_output = self.softmax(x, 1)
    
            return exp_x, exp_x_sum, softmax_output
    
    if __name__ == "__main__":
        model = CustomModel()
        model = model.to("cpu")
        model = model.eval()
    
        dummy_input = torch.rand((1, 4096, 4096), requires_grad=False)
    
        with torch.no_grad():
            _ = model(dummy_input)
    
            torch.onnx.export(model,
                              dummy_input,
                              "model.onnx",
                              export_params=True,
                              verbose=True,
                              do_constant_folding=True,
                              input_names=['input'],
                              output_names=['exp_x', 'exp_x_sum', 'softmax_output'])
    
    
        core = ov.Core()
        ov_model = ov.convert_model("model.onnx")
        compiled_model = core.compile_model(model=ov_model, device_name="GPU.0")
        
        np.random.seed(1111)
        openvino_dummy_input = np.random.rand(1, 4096, 4096)
    
        output_1 = compiled_model(openvino_dummy_input)
        output_2 = compiled_model(openvino_dummy_input)
    
        print(output_1[0])
        print(output_2[0])
        print(np.all(output_1[0] == output_2[0]))
    
        print(output_1[1])
        print(output_2[1])
        print(np.all(output_1[1] == output_2[1]))
    
        print(output_1[2])
        print(output_2[2])
        print(np.all(output_1[2] == output_2[2]))

And this is Relevant log output

> Exported graph: graph(%input : Float(1, 4096, 4096, strides=[16777216, 4096, 1], requires_grad=0, device=cpu)):
>   %exp_x : Float(1, 4096, 4096, strides=[16777216, 4096, 1], requires_grad=0, device=cpu) = onnx::Exp[onnx_name="/Exp"](%input), scope: __main__.CustomModel:: # /home/user/Desktop/openvino/bug_report/main.py:16:0
>   %onnx::ReduceSum_2 : Long(1, strides=[1], device=cpu) = onnx::Constant[value={1}]()
>   %exp_x_sum : Float(1, 1, 4096, strides=[4096, 4096, 1], requires_grad=0, device=cpu) = onnx::ReduceSum[keepdims=1, onnx_name="/ReduceSum"](%exp_x, %onnx::ReduceSum_2), scope: __main__.CustomModel:: # /home/user/Desktop/openvino/bug_report/main.py:18:0
>   %softmax_output : Float(1, 4096, 4096, strides=[16777216, 4096, 1], requires_grad=0, device=cpu) = onnx::Div[onnx_name="/Div"](%exp_x, %exp_x_sum), scope: __main__.CustomModel:: # /home/user/Desktop/openvino/bug_report/main.py:20:0
>   return (%exp_x, %exp_x_sum, %softmax_output)
> 
> [[[1.1002629 2.5218775 1.409977  ... 1.8604656 2.1252458 2.570416 ]
>   [1.546227  1.196379  1.3293308 ... 1.8934513 1.1088103 1.7279357]
>   [1.9836317 2.586512  1.1108454 ... 1.470562  2.2241154 1.0347315]
>   ...
>   [1.0100845 2.716463  2.6036747 ... 1.2251861 1.829708  1.9024576]
>   [2.5323563 1.7440348 1.3570391 ... 1.5804602 2.0116925 2.1194026]
>   [1.4737493 2.2389796 2.2041445 ... 1.9816811 1.7517757 1.1047524]]]
> [[[1.1002629 2.5218775 1.409977  ... 1.8604656 2.1252458 2.570416 ]
>   [1.546227  1.196379  1.3293308 ... 1.8934513 1.1088103 1.7279357]
>   [1.9836317 2.586512  1.1108454 ... 1.470562  2.2241154 1.0347315]
>   ...
>   [1.0100845 2.716463  2.6036747 ... 1.2251861 1.829708  1.9024576]
>   [2.5323563 1.7440348 1.3570391 ... 1.5804602 2.0116925 2.1194026]
>   [1.4737493 2.2389796 2.2041445 ... 1.9816811 1.7517757 1.1047524]]]
> True
> [[[7025.714  7041.125  7050.5376 ... 7035.315  7041.6514 7037.6953]]]
> [[[7025.714  7041.125  7050.5376 ... 7035.315  7041.6514 7037.6953]]]
> True
> [[[0.00014198 0.00017929 0.00017977 ... 0.00026369 0.00030136 0.0003643 ]
>   [0.00021935 0.00016963 0.00018847 ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   ...
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]]]
> [[[0.00019956 0.00019324 0.00014186 ... 0.00026369 0.00030136 0.0003643 ]
>   [0.00021935 0.00016963 0.00018847 ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   ...
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]
>   [0.         0.         0.         ... 0.         0.         0.        ]]]
> False

**Comment by DavitGrigoryan132** [comment]
@mvafin Hi, With directly passing pytorch model to `convert_model` works for me too, but I need to convert it from ONNX in my program.
I tried using `config={"INFERENCE_PRECISION_HINT": ov.Type.f32}` and that didn't help me

**Comment by slyalin** [comment]
Diff between onnx converted and pytorch directly converted models has (1) different axis input types for ReduceSum (i32 for PyTorch, i64 for ONNX), and (2) shapes are static for onnx and dynamic for pytorch. Looks like a bug in the GPU plugin.

**Comment by geunhwan** [comment]
Fix has been provided.
@DavitGrigoryan132 please verify.

**Comment by DavitGrigoryan132** [comment]
@geunhwan Yes it works, thanks
